### ISSUE TITLE
[heft] Enable Storybook to work with the Heft toolchain

### ISSUE THREAD

**Issue by octogonz** [microsoft/rushstack#2833]
## Summary

At my work, two teams recently both asked to use [Storybook](https://storybook.js.org/) for testing their React UI controls.  These projects are built using Heft.  When I tried the basic Storybook "getting started" with [heft-webpack-basic-tutorial](https://github.com/microsoft/rushstack-samples/tree/main/heft/heft-webpack-basic-tutorial) it did not work.

- Heft should include a sample project illustrating how to use Storybook
- I'm pretty sure we also need to wrap Storybook behind a `@rushstack/heft-storybook-plugin` (similar to the approach of `@rushstack/heft-jest-plugin`)

## Details
Storybook launches a localhost website displaying an interactive gallery of your UI controls.  This provides a nice isolated environment for developing the controls, and a way of testing them in isolate, and documenting them for app developers.

From my investigation:

- Storybook basically ignores your project's toolchain, and instead builds everything using a custom toolchain that includes lots of configs and transforms that they hope will be able to compile typical source files.  By "custom toolchain" I don't mean a Webpack config, but actually an [elaborate command-line tool](https://github.com/storybookjs/storybook/blob/8bca9ec15339fd6b471ba7ff0fd23b02b0717abd/lib/core-server/src/build-dev.ts#L26) that spans multiple NPM packages (and fortunately has a nice API interface we can use instead of the CLI).

- But they do look at your project's `node_modules` folder and, for example, it will choose Webpack 4 vs 5 based on what they find there. The toolchain dependencies get installed into your project's `node_modules` folder, naively assuming that the Storybook toolchain won't have any conflicts with the project toolchain.  

- There are actually two separate Webpack configs:  (1) the `manager` config runs the Storybook app and the entirety of its UI surfaces, and (2) the `preview` config builds our React controls, which are isolated in an HTML iframe.  This iframe loads its own build of the Storybook runtime bundle and vendor bundle. I'm guessing the purpose is to prevent the Storybook page from crashing and losing state if your project code has a syntax error.

- The Storybook web app and its add-ons are distributed as NPM packages that are mostly prebuilt.  But interestingly they still need Webpack to link everything together.  (And many vendor libraries like Lodash are unbundled, so the Webpack stats dump shows literally thousands of individual file paths getting processed.)

In order to make Storybook work great with Heft, there's basically 3 problems we can solve:

1. **Integration Problem:** Storybook's toolchain is geared for starter projects and is likely to have trouble compiling code that relies on Heft preprocessors. For example, it failed on a `const enum` in **heft-webpack-basic-tutorial** project, because they are trying to use isolated modules mode.

   **Proposed Solution:** Point Storybook at our `lib/*.js` output folder (instead of `src/*.tsx`), and make a simple Heft plugin that launches Storybook alongside Heft's watch mode. (As an experiment, I tried invoking them manually, and works just fine.)  This is essentially the same approach as what `heft-webpack-plugin` does.

2. **Integration Problem:** Storybook builds pretty slowly.  My trivial project took about ~40 seconds to bundle because it's babelling everything including all the vendor libraries.  I suspect some of this is redundant, since our `lib/*.js` folder is already fully transpiled.
 
   **Proposed Solution:** We could try making a Storybook preset that trims down the configuration. It might even be possible to eliminate Babel entirely. (Heft transpiles entirely via `tsc` and does not load Babel at all.)

3. **Integration Problem:** Storybook adds a bunch of its NPM packages to the project's **package.json** file, which is problematic. (If you're going to operate your own build environment as a parallel universe, you can't expect your dependency versions to be compatible with whatever the project is doing.)  Also, in a monorepo you want to avoid copy+pasting dependencies into 20 different project folders, which we do using [rig packages](https://rushstack.io/pages/heft/rig_packages/).

   **Proposed Solution:** If we make a `heft-storybook-plugin`, I feel like it can bring along the Storybook engine and the most common dependencies. This is how `heft-jest-plugin` works.

BTW in our investigation, @lmatteis-wm found that someone had started work on a [@hulkstack/heft-plugin-storybook](https://npm.io/package/@hulkstack/heft-plugin-storybook) but later unpublished it for some reason. We found the tarball, and it seems like basically the same idea of invoking Storybook in parallel with `heft build --watch`.

**Comment by sdalonzo** [comment]
My team is very interested in this work and would be happy to contribute to design, implementation and/or testing as needed.

**Current state** 

- We're still in the process of moving to Heft
- Currently completing Storybook 6 migration (intended to delay until after a Heft plugin existed but v5 has grown too brittle and time consuming to maintain)
- Our current Storybook 6 implementation is fairly vanilla to start, which should serve as a good starting point for a Heft+Storybook integration
- Many React component projects and a few apps have Storybook and our internal preset in them (in Storybook v5, anyway, we were able to use the preset like a Heft plugin as far as providing its own dependencies)
- In addition to that, we have a "Universal Storybook" project
  - current implementation uses Rush libs to find all projects with Storybook as a dependency and constructs and array of globs
  - build and UI are slow due to sheer volume of stories
  - not incremental
  - is built on PR and deployed for design/product review "storybook sandbox"
  - also deployed a more canonical place on changes to master

**Ambitions with v6 and Heft**

- [Storybook Composition](https://storybook.js.org/docs/react/workflows/storybook-composition)
  - In this model, you build the storybook for each project discretely and you mount N storybooks in a shell storybook
  - In my mental model of using Storybook with Heft, Heft builds the storybook (`build-storybook`). Then, in CI, "leaf" storybooks are built incrementally, but a complete "Universal Storybook" can be composed. Maybe to a separate cache? bulk command with `incremental: true`?
- Plugin flexibility per project - there are a lot of really interesting but very specific plugins for SBv6 (ie mocking apollo client, next-router for example). We'd like to include a core set with every storybook (essentials, some custom viewports, some themes for whitelabel partners, etc) without the overhead of including all of the potentially useful ones in every project.
- The Docs page (now baked into addon-essentials) is really useful in terms of generating a props table based on prop-types declarations (admittedly, we are early in our TypeScript journey). I'd be loathe to give that up, as it has value to multiple stakeholders.
- Potential Chromatic integration: [Chromatic](https://www.chromatic.com/) is an excellent tool built by the same folks who build and maintain storybook. Their platform hosts storybooks for you and provides visual regression testing/change approval process and some other nice to haves. I was considering invoking their APIs from some code that would use the recently published Change Analyzer to figure out which storybooks to build and sent to Chromatic.

Comments on proposed problems:
1. Integration problem - I completely agree with the proposed, and leaning into Heft's strengths.
2. I suspect you are correct here. You can override much of the webpack config in storybook 6 as part of a [preset](https://storybook.js.org/docs/react/addons/writing-presets#webpack). I feel like build cache has an important role to play here, as related to the composition concept. Could their conceptually be a storybook cache separate from the regular build cache?
3. `heft-storybook-plugin` seems like a great approach to me.

**Comment by octogonz** [comment]
Here is my PR with an initial implementation of `heft-storybook-plugin`: https://github.com/microsoft/rushstack/pull/2885

@sdalonzo thanks for your note, lemme read over it and follow up with feedback.

**Comment by octogonz** [comment]
@sdalonzo Thanks for your writeup. It is a nice complement, providing a "how we want to use Storybook", whereas my PR is more about "mechanics of invoking Storybook in a Heft project."  So your requests look like good next steps once the plugin is released.

I should clarify that PR #2885 doesn't tackle all three "Integration Problems" identified above:

- **Problem 1 (processing Heft with Heft preprocessors)**: Storybook's inputs are `lib/*.js` files produced by Heft, so its toolchain doesn't need to know anything about Heft's TypeScript configuration or preprocessors.
- **Problem 2 (Storybook builds pretty slowly because it's doing extraneous work)**: Not tackled yet.  `@storybook/react` builds in its normal default way.
- **Problem 3 (separating the NPM dependencies)**: This is solved by the "storykit package" concept (as described [here](https://github.com/microsoft/rushstack/blob/octogonz/heft-storybook/heft-plugins/heft-storybook-plugin/src/StorybookPlugin.ts#L54)).

Now some specific replies:

> Currently completing Storybook 6 migration

In its current form, the Heft plugin has no dependencies on Storybook NPM packages.  They are obtained from the "storykit package" which you define yourself.  Thus it's likely to work with any Storybook verison.

> - build and UI are slow due to sheer volume of stories
> - not incremental

This is an open question.  My PR implements watch mode, but does not wire up `heft build` to produce a Storybook bundle, nor does it consider where such a bundle should be written so that Rush can cache it.  Based on your description above, that sounds like the most important next step to tackle.

> We'd like to include a core set with every storybook (essentials, some custom viewports, some themes for whitelabel partners, etc) without the overhead of including all of the potentially useful ones in every project.

I think we would solve that by wiring up `.storybook/main.js` to inherit settings from a [rig package](https://rushstack.io/pages/heft/rig_packages/). That seems relatively easy to implement.  The `"storykitPackageName"` and `startupModulePath` settings are already riggable.
